### PR TITLE
[SPARK-36465][SS] Dynamic gap duration in session window

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch supports dynamic gap duration in session window.

### Why are the changes needed?

The gap duration used in session window for now is a static value. To support more complex usage, it is better to support dynamic gap duration which determines the gap duration by looking at the current data. For example, in our usecase, we may have different gap by looking at the certain column in the input rows.

### Does this PR introduce _any_ user-facing change?

Yes, users can specify dynamic gap duration.

### How was this patch tested?

Modified existing tests and new test.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Thanks for the contribution!

For the functionality of dynamic gap duration in session window, there are some proposals in other frameworks and I think these proposals make sense despite the difference - these proposals are based on non-SQL API whereas this is based on SQL API.

https://ci.apache.org/projects/flink/flink-docs-release-1.13/docs/dev/datastream/operators/windows/
https://cwiki.apache.org/confluence/display/KAFKA/KIP-362%3A+Support+dynamic+gap+session+window

Basically end users should be able to apply some logic against the row to extract the gap duration, and we should ensure the usage is simple enough. For example, suppose a use case we apply gap duration to 30 minutes for working hours (based on event time column) and 10 minutes for non-working hours. (It could also be based on event type column or something like.) How to represent this query?

**Comment by HeartSaVioR** [comment]
It would be nice to have some test cases / examples if we can deal with such case with UDF / SQL expression.

**Comment by viirya** [comment]
Thanks @HeartSaVioR ! Yea, let me add more examples/tests with some UDF / SQL expressions.

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
retest this please

**Comment by HeartSaVioR** [comment]
Let's handle a corner case; looks like CalendarInterval can be negative which we should not allow here. "0" gap interval might be tricky case so if there's no specific use case it might be better to avoid it.

Also we may need to explain that the session's end time is NOT the event time + gap duration of the last event when dynamic gap duration is used. This is different from static gap duration.

**Comment by HeartSaVioR** [comment]
And let's update the SS guide doc to include the example on dynamic gap of session window, with NOTEs I mentioned above.

**Comment by viirya** [comment]
> Let's handle a corner case; looks like CalendarInterval can be negative which we should not allow here. "0" gap interval might be tricky case so if there's no specific use case it might be better to avoid it.
>  (We probably need to filter out these events instead of failing the query. Not sure we need to log each event though. Probably another metric like dropped rows?)

Hmm, filtering out the events with negative/zero gap duration is an option. It should be as straightforward as adding a `Filter` with predicate on session start/end times.

`Filter` will be shown in UI with metric, although it doesn't log specially for the events dropped.

If logging such event or separate metric is needed, instead of `Filter`, we may need to manually filter out the events in the first operator processing the events. It means more code change is necessary.

**Comment by viirya** [comment]
> And let's update the SS guide doc to include the example on dynamic gap of session window, with NOTEs I mentioned above.

Sure.

**Comment by viirya** [comment]
And, of course, we need to clearly document that negative/zero gap duration events will be filtered out.

**Comment by HeartSaVioR** [comment]
End users are not easily indicated about the relation between session window aggregation and filter. It'd be better if we handle it along with dropping late rows.

Btw I found a missing spot in session window, we don't update the metric for dropped late rows. I'll fix it quickly.

**Comment by viirya** [comment]
> Also we may need to explain that the session's end time is NOT the event time + gap duration of the last event when dynamic gap duration is used. This is different from static gap duration.

I don't see in the SS guide we mention that the session's end time is the event time + gap duration of the last event? Do you mean this sentence?

> A session window closes when there's no input received within gap duration after receiving the latest input.

**Comment by HeartSaVioR** [comment]
I realized my bad after looking into your change... sorry about that. It doesn't make sense to filter out from all endpoints, as we will end up with touching everywhere which may feel odd. 

Let's just do what we've done with time window... We are no longer counting the case, but I think even we want to count the case when filtering out, we need a new local/physical plan to inject instead of injecting filter, instead of touching all endpoints.
https://github.com/apache/spark/blob/f7c85b855ba99757c750dd0a2f7aced788c89374/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3910-L3921

Sorry again for being back and forth.

**Comment by HeartSaVioR** [comment]
> I don't see in the SS guide we mention that the session's end time is the event time + gap duration of the last event? Do you mean this sentence?
>> A session window closes when there's no input received within gap duration after receiving the latest input.
> It looks like not saying that the session's end time is the event time + gap duration of the last event.

The sentence only assumes static gap duration, and it's no longer true for dynamic gap duration. At least current logic won't work like so.

We are technically "merging" sessions, which means the session's range is union of all merged sessions' range. It's simply possible that `earlier event + gap duration` is later than `latest event + gap duration`, and in this case the session's end time is `earlier event + gap duration` instead of `latest event + gap duration`.

We probably need to explain the dynamic gap duration as a separate figure/sentence, so that end users can understand it is working as "merging" sessions, or if we don't like to expose technical aspect, union of sessions' range.

**Comment by HeartSaVioR** [comment]
(Actually it is very hard to implement dynamic gap duration with respecting the sentence "A session window closes when there's no input received within gap duration after receiving the latest input.". Due to the nature of event time processing, the session can be "shorter" after applying some events into session, and it can trigger "split" the session which means we no longer able to aggregate in prior.)

File: docs/structured-streaming-programming-guide.md [hunk 1]
```diff
@@ -1134,6 +1134,73 @@ sessionizedCounts = events \
...
+{% endhighlight %}
+
+</div>
+<div data-lang="python"  markdown="1">
+{% highlight python %}
+from pyspark.sql import functions as F
+
+events = ...  # streaming DataFrame of schema { timestamp: Timestamp, userId: String }
+
+session_window = session_window(events.timestamp, \
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:1188]: Does it work? We check the type of gapDuration and expect str type here.

https://github.com/apache/spark/blob/f7c85b855ba99757c750dd0a2f7aced788c89374/python/pyspark/sql/functions.py#L2336-L2368

https://github.com/apache/spark/blob/bbf988bd73b00d18dd1d443f225b3915a2c4433f/python/pyspark/sql/functions.pyi#L139

File: docs/structured-streaming-programming-guide.md [hunk 2]
```diff
@@ -1134,6 +1134,73 @@ sessionizedCounts = events \
...
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+import spark.implicits._
+
+val events = ... // streaming DataFrame of schema { timestamp: Timestamp, userId: String }
+
+val sessionWindow = SessionWindow($"timestamp".expr, when($"userId" === "user1", "5 seconds")
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:1149]: Why not adding override method of `session_window` into functions.scala? It sounds inconvenient if I need to remember dynamic gap duration is only supported with `SessionWindow` expression.

File: docs/structured-streaming-programming-guide.md [hunk 3]
```diff
@@ -1134,6 +1134,73 @@ sessionizedCounts = events \
...
+  .when($"userId" === "user2", "20 seconds")
+  .otherwise("5 minutes")
```
Comment (HeartSaVioR) [inline docs/structured-streaming-programming-guide.md:1151]: nit: Probably missed `)` at the last.

**Review by HeartSaVioR — COMMENTED** [review]
I'm wondering whether we can still leverage `session_window` function to deal with dynamic gap duration as well. If that is not possible or hurt usability of static gap duration, we may need to take other function name.

File: docs/structured-streaming-programming-guide.md (hunk 2 above)
Comment (viirya) [reply docs/structured-streaming-programming-guide.md:1149]: Yea, also had the same thought. Let me add another `session_window` function.

File: docs/structured-streaming-programming-guide.md (hunk 3 above)
Comment (viirya) [reply docs/structured-streaming-programming-guide.md:1151]: Oops!

File: docs/structured-streaming-programming-guide.md (hunk 1 above)
Comment (viirya) [reply docs/structured-streaming-programming-guide.md:1188]: Oh, I don't realize python `session_window` doesn't accept an expression...

**Comment by viirya** [comment]
> We are technically "merging" sessions, which means the session's range is union of all merged sessions' range. It's simply possible that `earlier event + gap duration` is later than `latest event + gap duration`, and in this case the session's end time is `earlier event + gap duration` instead of `latest event + gap duration`.
> 
> We probably need to explain the dynamic gap duration as a separate figure/sentence, so that end users can understand it is working as "merging" sessions, or if we don't like to expose technical aspect, union of sessions' range.

I prefer not to expose too much technical aspect.

**Comment by viirya** [comment]
> Due to the nature of event time processing, the session can be "shorter" after applying some events into session, and it can trigger "split" the session which means we no longer able to aggregate in prior.)

Not sure I follow it up correctly. If the session is the union of union of events' range, how can it be shorter after applying (unionning?) more events? We don't have negative gap duration.

**Comment by HeartSaVioR** [comment]
> Not sure I follow it up correctly. If the session is the union of union of events' range, how can it be shorter after applying (unionning?) more events? We don't have negative gap duration.

No I meant it can happen if we are defining the session end as "time + gap duration of the last event" instead of defining the session end as "max of events' time + gap duration". Events can be out of order.

**Comment by HeartSaVioR** [comment]
There's a tricky case for the definition of session end as "time + gap duration of the last event". Suppose the events are coming like below:

* e1: time 15, gap duration 15
* e2: time 25, gap duration 10
* e3: time 20, gap duration 2

e1 and e2 are mergeable so they will compose a session S1. e3 comes, and also included to the S1, but when we reevaluate events in session S1, e2 should NOT be included in S1. S1 can include e1 and e3, and e2 should compose a new session.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingSessionWindowSuite.scala [hunk 4]
```diff
@@ -256,6 +256,102 @@ class StreamingSessionWindowSuite extends StreamTest
...
+      CheckNewAnswer(
+      ),
+
+      // concatenating multiple previous sessions into one
+      AddData(inputData, ("spark streaming", 30L)),
+      // watermark: 11
+      // current sessions
+      // ("spark", 25, 50, 25, 3),
+      // ("streaming", 25, 51, 26, 4),
+      // ("hello", 40, 51, 11, 2),
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingSessionWindowSuite.scala:323]: `("hello", 40, 42, 2, 2)`

**Review by HeartSaVioR — COMMENTED** [review]
Looks good in overall. Only minors.

**Comment by viirya** [comment]
Thanks @HeartSaVioR ! Updated.

**Comment by HeartSaVioR** [comment]
There's a voice about having overloaded function by parameter type on SQL function; please allow us having several days to audit the change in point of API view. Thanks for understanding!

**Comment by HyukjinKwon** [comment]
I think it's fine to go ahead.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master/branch-3.2.

**Comment by viirya** [comment]
Thanks @HeartSaVioR @HyukjinKwon !
